great_expectations/data_context/types/base.py

```diff
@@ -2269,82 +2296,308 @@ def get_updated_action_list(
         base_action_list_dict = {action["name"]: action for action in base_action_list}
+
         for other_action in other_action_list:
             other_action_name = other_action["name"]
             if other_action_name in base_action_list_dict:
-                if other_action["action"] is None:
-                    base_action_list_dict.pop(other_action_name)
-                else:
+                if other_action["action"]:
                     nested_update(
                         base_action_list_dict[other_action_name],
                         other_action,
                         dedup=True,
                     )
+                else:
+                    base_action_list_dict.pop(other_action_name)
             else:
                 base_action_list_dict[other_action_name] = other_action
+
+        for other_action in other_action_list:
+            other_action_name = other_action["name"]
+            if other_action_name in base_action_list_dict:
+                if not other_action["action"]:
+                    base_action_list_dict.pop(other_action_name)
+
         return list(base_action_list_dict.values())
 
     @property
-    def evaluation_parameters(self):
+    def evaluation_parameters(self) -> dict:
         return self._evaluation_parameters
 
+    @evaluation_parameters.setter
+    def evaluation_parameters(self, value: dict):
+        self._evaluation_parameters = value
+
     @property
-    def runtime_configuration(self):
+    def runtime_configuration(self) -> dict:
         return self._runtime_configuration
 
-    def __repr__(self):
-        batch_data_list = []
-        if len(self.validations) > 0:
-            for val in self.validations:
-                if (val["batch_request"].get("runtime_parameters") is not None) and (
-                    val["batch_request"]["runtime_parameters"].get("batch_data")
+    @runtime_configuration.setter
+    def runtime_configuration(self, value: dict):
+        self._runtime_configuration = value
+
+    def __deepcopy__(self, memo):
+        batch_data_references: Tuple[
+            Optional[Any], Optional[List[Any]]
+        ] = get_runtime_parameters_batch_data_references_from_config(config=self)
+        delete_runtime_parameters_batch_data_references_from_config(config=self)
+
+        cls = self.__class__
+        result = cls.__new__(cls)
+        result._commented_map = CommentedMap()
+
+        memo[id(self)] = result
+        for key, value in self.to_json_dict().items():
+            # noinspection PyArgumentList
+            value_copy = copy.deepcopy(value, memo)
+            setattr(result, key, value_copy)
+
+        restore_runtime_parameters_batch_data_references_into_config(
+            config=self, batch_data_references=batch_data_references
+        )
+        restore_runtime_parameters_batch_data_references_into_config(
+            config=result, batch_data_references=batch_data_references
+        )
+
+        return result
```

Comment:
Can a custom `__deepcopy__` not be avoided here? Why isn't this something we can do where required, like we are doing in other places using these new functions in this PR?

Reply:
@NathanFarmer This custom `__deepcoppy__` enables us to simplify / delete a lot of code.  It fits, because it "knows" how to handle the copying for the (complex) `CheckpointConfig` object.